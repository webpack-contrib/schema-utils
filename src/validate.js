import Ajv from 'ajv';
import ajvKeywords from 'ajv-keywords';

import absolutePathKeyword from './keywords/absolutePath';

import ValidationError from './ValidationError';

const ajv = new Ajv({
  allErrors: true,
  verbose: true,
});

ajvKeywords(ajv, [
  'instanceof',
  'formatMinimum',
  'formatMaximum',
  'patternRequired',
]);

// Custom keywords
absolutePathKeyword(ajv);

function validate(schema, options, configuration = {}) {
  let errors = [];

  if (Array.isArray(options)) {
    errors = options.map((nestedOptions) =>
      validateObject(schema, nestedOptions)
    );

    errors.forEach((list, idx) => {
      const applyPrefix = (error) => {
        // eslint-disable-next-line no-param-reassign
        error.dataPath = `[${idx}]${error.dataPath}`;

        if (error.children) {
          error.children.forEach(applyPrefix);
        }
      };

      list.forEach(applyPrefix);
    });

    errors = errors.reduce((arr, items) => arr.concat(items), []);
  } else {
    errors = validateObject(schema, options);
  }

  if (errors.length > 0) {
    throw new ValidationError(errors, schema, configuration);
  }

  return errors;
}

function validateObject(schema, options) {
  const compiledSchema = ajv.compile(schema);
  const valid = compiledSchema(options);

  return valid ? [] : filterErrors(compiledSchema.errors);
}

function filterErrors(errors) {
  let newErrors = [];

  for (const error of errors) {
    const { dataPath } = error;
    let children = [];

    newErrors = newErrors.filter((oldError) => {
      if (oldError.dataPath.includes(dataPath)) {
        if (oldError.children) {
          children = children.concat(oldError.children.slice(0));
        }

        // eslint-disable-next-line no-undefined, no-param-reassign
        oldError.children = undefined;
        children.push(oldError);

        return false;
      }

      return true;
    });

    if (children.length) {
      error.children = children;
    }

    newErrors.push(error);
  }

  return newErrors;
}

export default validate;
