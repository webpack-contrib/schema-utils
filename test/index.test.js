import validate from '../src';

test('Valid', () => {
  const name = 'TestLoader';
  const schema = 'test/fixtures/schema.json';
  const target = {
    type() {},
    array: ['a'],
    string: 'hello',
    object: {
      prop: false,
      object: {
        prop: false,
      },
    },
    boolean: true,
    instance: new RegExp(''),
  };
  const result = validate({ name, schema, target });

  expect(result).toBe(true);
});

describe('Error', () => {
  const name = 'TestLoader';
  const schema = 'test/fixtures/schema.json';
  const target = {
    type: null,
    array: {},
    string: false,
    object: {
      prop: 1,
      object: {
        prop: 1,
      },
    },
    boolean: 'hello',
    instance() {},
  };

  const doit = () => validate({ name, schema, target });

  test('should throw error', () => {
    expect(doit).toThrowError();
    expect(doit).toThrowErrorMatchingSnapshot();
  });

  test('should have errors for every option key', () => {
    try {
      doit();
    } catch (err) {
      const errors = err.meta.errors.map((error) => error.dataPath);

      const expected = [
        '.array',
        '.boolean',
        '.instance',
        '.object.object.prop',
        '.object.prop',
        '.string',
        '.type',
      ];

      expect(errors).toMatchObject(expected);
      expect(err.meta.errors).toMatchSnapshot();
    }
  });
});
