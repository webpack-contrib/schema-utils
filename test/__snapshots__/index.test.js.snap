// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error should have errors for every option key 1`] = `
Array [
  Object {
    "dataPath": ".array",
    "keyword": "type",
    "message": "should be array",
    "params": Object {
      "type": "array",
    },
    "schemaPath": "#/properties/array/type",
  },
  Object {
    "dataPath": ".boolean",
    "keyword": "type",
    "message": "should be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/properties/boolean/type",
  },
  Object {
    "dataPath": ".instance",
    "keyword": "instanceof",
    "message": "should pass \\"instanceof\\" keyword validation",
    "params": Object {
      "keyword": "instanceof",
    },
    "schemaPath": "#/properties/instance/instanceof",
  },
  Object {
    "dataPath": ".object.object.prop",
    "keyword": "type",
    "message": "should be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/properties/object/properties/object/properties/prop/type",
  },
  Object {
    "dataPath": ".object.prop",
    "keyword": "type",
    "message": "should be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/properties/object/properties/prop/type",
  },
  Object {
    "dataPath": ".string",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/string/type",
  },
  Object {
    "dataPath": ".type",
    "keyword": "typeof",
    "message": "should pass \\"typeof\\" keyword validation",
    "params": Object {
      "keyword": "typeof",
    },
    "schemaPath": "#/properties/type/typeof",
  },
]
`;

exports[`Error should throw error 1`] = `
"TestLoader

  [4mOptions Validation Error[24m

  [2moptions[22m.array               [34mshould be array[39m
  [2moptions[22m.boolean             [34mshould be boolean[39m
  [2moptions[22m.instance            [34mshould pass \\"instanceof\\" keyword validation[39m
  [2moptions[22m.object.object.prop  [34mshould be boolean[39m
  [2moptions[22m.object.prop         [34mshould be boolean[39m
  [2moptions[22m.string              [34mshould be string[39m
  [2moptions[22m.type                [34mshould pass \\"typeof\\" keyword validation[39m
"
`;
